topology/probes/k8s/node.go

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"sync"
+
+	"github.com/skydive-project/skydive/topology/graph"
+
+	"k8s.io/api/core/v1"
+)
+
+type nodeCache struct {
+	sync.RWMutex
+	defaultKubeCacheEventHandler
+	graph.DefaultGraphListener
+	*kubeCache
+	graph *graph.Graph
+}
+
+func (c *nodeCache) getMetadata(node *v1.Node) graph.Metadata {
+	return graph.Metadata{
+		"Type":    "host",
+		"Manager": "k8s",
+		"Name":    node.GetName(),
+		"K8s":     node,
+	}
+}
+
+func (c *nodeCache) OnAdd(obj interface{}) {
+	c.OnUpdate(obj, obj)
+}
+
+func (c *nodeCache) OnUpdate(old, new interface{}) {
+	newNode := new.(*v1.Node)
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.graph.Lock()
+	defer c.graph.Unlock()
+
+	graphMeta := c.getMetadata(newNode)
+	hostIndexer := graph.NewMetadataIndexer(c.graph, graph.Metadata{"Type": "host"}, "Name")
+	hostNodes := hostIndexer.Get(newNode.GetName())
+	if len(hostNodes) == 0 {
+		c.graph.NewNode(graph.Identifier(newNode.GetUID()), graphMeta)
+	} else {
+		if val, ok := graphMeta["Manager"]; ok && val == "k8s" {
+			c.graph.SetMetadata(hostNodes[0], graphMeta)
```

Comment:
SetMetadata overrides all the existing metadata. So if the user added custom metadata through the configuration files, they will be overriden. You could either use AddMetadata, or get the host metadata, add the fields you want and use SetMetadata

Reply:
ok in next update I will use `AddMetadata()` to only update:
- K8s: which retains the raw metadata provided by k8s objects
- Manager: "k8s": so that we get he icon indicating this is tracked by k8s